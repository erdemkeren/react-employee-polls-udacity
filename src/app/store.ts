import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../features/auth/authSlice';
import pollReducer from '../features/poll/pollSlice';
import userReducer from '../features/user/userSlice';

export const store = configureStore({
  reducer: {
    auth: authReducer,
    poll: pollReducer,
    user: userReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
