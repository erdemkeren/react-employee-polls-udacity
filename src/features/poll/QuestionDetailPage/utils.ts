import {Question} from "../../../types";
import {QuestionStats} from "./types";

const getPercentage = (total: number, val: number): number => Math.round(100 / total * val);

export const getStatsForQuestion = (question: Question): QuestionStats => {
    const oneN = question.optionOne.votes.length;
    const twoN = question.optionTwo.votes.length
    const total = oneN + twoN;
    const oneP = getPercentage(total, oneN);
    const twoP = getPercentage(total, twoN);

    return {oneN, twoN, oneP, twoP};
};

export const getValueForQuestion = (question: Question, userId: string): string | null => {
    if (question.optionOne.votes.includes(userId)) {
        return 'optionOne';
    }

    if (question.optionTwo.votes.includes(userId)) {
        return 'optionTwo';
    }

    return null;
}
