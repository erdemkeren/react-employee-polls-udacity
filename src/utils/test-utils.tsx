import React, { ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import authReducer from '../features/auth/authSlice';
import pollSlice from '../features/poll/pollSlice';
import userSlice from '../features/user/userSlice';

function render(
    ui: React.ReactElement,
    {
        store = configureStore({
            reducer: {
                auth: authReducer,
                poll: pollSlice,
                user: userSlice,
            },
        }),
        ...renderOptions
    } = {}
) {
    type Props = { children?: ReactNode };
    const Wrapper: React.FC = ({ children }: Props) => (
        <Provider store={store}>{children}</Provider>
    )
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

export * from '@testing-library/react';
export { render }
