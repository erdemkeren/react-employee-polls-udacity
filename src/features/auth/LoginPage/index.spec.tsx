import Login from "./";
import { render, screen, fireEvent } from '../../../utils/test-utils';
import * as hooks from '../../../app/hooks';

describe('Login form', function () {
    it('should render', function () {
        const { container } = render(<Login />);

        expect(container).toMatchSnapshot();
    });

    it('should trigger login', function () {
        const useAppDispatch = jest.spyOn(hooks, 'useAppDispatch')
        const useAppSelector = jest.spyOn(hooks, 'useAppSelector')

        useAppSelector.mockReturnValue({
            isAuthenticated: false,
            userId: undefined,
            status: 'idle',
        });

        const dummyDispatch = jest.fn();
        useAppDispatch.mockReturnValue(dummyDispatch);

        render(<Login />);

        const button = screen.getByRole('button', /Sign In/i)
        fireEvent.click(button);
        expect(dummyDispatch).toBeCalledTimes(1);

        useAppDispatch.mockRestore();
        useAppSelector.mockRestore();
    });

    it('should disable the login button', function () {
        render(<Login />);

        const button = screen.getByRole('button', /Sign In/i)
        fireEvent.click(button);
        expect(button).toBeDisabled();
    });

    it('should show failed login info', function () {
        const useAppDispatch = jest.spyOn(hooks, 'useAppDispatch')
        const useAppSelector = jest.spyOn(hooks, 'useAppSelector')

        useAppSelector.mockReturnValue({
            isAuthenticated: false,
            userId: undefined,
            status: 'idle',
        });

        useAppSelector.mockReturnValue({
            status: 'failed',
        })

        const { container } = render(<Login />);

        expect(container).toHaveTextContent('The authentication was not successful.');

        expect(container).toMatchSnapshot();

        useAppDispatch.mockRestore();
        useAppSelector.mockRestore();
    });
});
