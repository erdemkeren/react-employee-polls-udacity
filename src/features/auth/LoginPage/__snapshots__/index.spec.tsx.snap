// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form should render 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-binzgt"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1999axt-MuiAvatar-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LockOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
          />
        </svg>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Sign in
      </h1>
      <form
        class="MuiBox-root css-164r41r"
        novalidate=""
      >
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused Mui-required css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="username"
            id="username-label"
          >
            Username
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="username"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="username"
              name="username"
              required=""
              type="text"
              value="erdemkeren"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1z7n62"
              >
                <span>
                  Username
                   *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="password"
            id="password-label"
          >
            Password
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="current-password"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="password"
              name="password"
              required=""
              type="password"
              value="pwd2022"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1z7n62"
              >
                <span>
                  Password
                   *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root css-1qudo6r-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
          tabindex="0"
          type="submit"
        >
          Sign In
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter css-ijul54-MuiTypography-root"
    >
      Copyright © 
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1kgfegd-MuiTypography-root-MuiLink-root"
        href="https://erdemkeren.com/"
      >
        Hilmi Erdem Keren
      </a>
       
      2021
      .
    </p>
  </main>
</div>
`;

exports[`Login form should show failed login info 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-binzgt"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1999axt-MuiAvatar-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LockOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
          />
        </svg>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Sign in
      </h1>
      <form
        class="MuiBox-root css-164r41r"
        novalidate=""
      >
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused Mui-required css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="username"
            id="username-label"
          >
            Username
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="username"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="username"
              name="username"
              required=""
              type="text"
              value="erdemkeren"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1z7n62"
              >
                <span>
                  Username
                   *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="password"
            id="password-label"
          >
            Password
            <span
              aria-hidden="true"
              class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="current-password"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="password"
              name="password"
              required=""
              type="password"
              value="pwd2022"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1z7n62"
              >
                <span>
                  Password
                   *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardWarning MuiAlert-standard css-rob5aa-MuiPaper-root-MuiAlert-root"
          role="alert"
        >
          <div
            class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="ReportProblemOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message css-acap47-MuiAlert-message"
          >
            The authentication was not successful.
          </div>
        </div>
        <button
          class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root css-1qudo6r-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
          tabindex="0"
          type="submit"
        >
          Sign In
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter css-ijul54-MuiTypography-root"
    >
      Copyright © 
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1kgfegd-MuiTypography-root-MuiLink-root"
        href="https://erdemkeren.com/"
      >
        Hilmi Erdem Keren
      </a>
       
      2021
      .
    </p>
  </main>
</div>
`;
